revive idba_ud

bernt-matthias opened the pull request:
One of our users needs idba_ud .. so I use the chance to revive the galaxy tool. The commit is based on https://toolshed.g2.bx.psu.edu/repos/stheil/idba_ud_wrapper (Revision: 2:edda44d54b59). Currently I only switched to conda, added test, help, and citation. 

Discussion: 
- our users needs a patched version of ibda_ud (she needed to change one of the hard coded limits .. I remember that it was the length of 128 that separated short and long reads). Do you see a way to realize this in conda/Galaxy
- do we want/need to keep the fq2fa tool (for interleaving and N removal)?

bgruening commented on tools/idba_ud/fq2fa.xml:
```diff
@@ -0,0 +1,55 @@
+<tool id="fq2fa" name="fq2fa" version="1.1.1 (Oct 18, 2012)">
```
please remove the data from the version

bgruening commented on tools/idba_ud/fq2fa.xml:
```diff
@@ -0,0 +1,55 @@
+<tool id="fq2fa" name="fq2fa" version="1.1.1 (Oct 18, 2012)">
+  <description>fq2fa is a simple tool given by IDBA package that allows to convert fastq to fasta, merge paired-end reads and filter out reads containing 'N's</description>
+  <command>
```
please add a requirement

bgruening commented on tools/idba_ud/fq2fa.xml:
```diff
@@ -0,0 +1,55 @@
+  <command>
+    fq2fa
+    #if $fileLayout.layout=='paired'
+      --paired $fileLayout.sequence
```
data params and text params should be single quoted

bgruening commented on tools/idba_ud/fq2fa.xml:
```diff
@@ -0,0 +1,55 @@
+      $fileLayout.sequence
+    #end if
+    $filter
+    $output
```
single quotes

bgruening commented on tools/idba_ud/fq2fa.xml:
```diff
@@ -0,0 +1,55 @@
+        <param name="sequence" type="data" format="fastq" label="Interleaved sequence file"/>
+      </when>
+      <when value="--merge">
+        <param name="sequenceR1" type="data" format="fastq" label="Forward sequence file (R1)"/>
```
only fastq or also fastq.gz?

bgruening commented on tools/idba_ud/fq2fa.xml:
```diff
@@ -0,0 +1,55 @@
+    <param name="filter" type="boolean" checked="false" truevalue="--filter" falsevalue="" label="Filter out reads containing 'N'"/>
+  </inputs>
+  <outputs>
+		<data name="output" format="fasta" label="Fasta from fq2fa"/>
```
please adjust the indentation ... 4 spaces like python

bgruening commented on tools/idba_ud/idba_ud.xml:
```diff
@@ -0,0 +1,99 @@
+
+	; 
+
+	mv out/scaffold.fa $output
```
quotes

bgruening commented on tools/idba_ud/idba_ud.xml:
```diff
@@ -0,0 +1,99 @@
+		<param name="prefix" type="integer" value="3" label="Prefix length used to build sub k-mer table (--prefix)"/>
+		<param name="min_count" type="integer" value="2" label="Minimum multiplicity for filtering k-mer when building the graph (--min_count)"/>
+		<param name="min_support" type="integer" value="1" label="Minimum supoort in each iteration (--min_support)"/>
+		<param name="num_threads" type="integer" value="0" label="Number of threads (--num_threads)"/>
```
you could use arguments= here everywhere ...

nsoranzo commented:
> do we want/need to keep the fq2fa tool (for interleaving and N removal)?

Not sure it is worth to have, since we have https://toolshed.g2.bx.psu.edu/view/devteam/fastq_to_fasta

bernt-matthias commented:
But: can it also do the interleaving in the same way?

nsoranzo commented:
No, for that I'd need to use a separate tool like  https://toolshed.g2.bx.psu.edu/view/devteam/fastq_paired_end_interlacer

bgruening commented on tools/idba_ud/idba_ud.xml:
```diff
@@ -0,0 +1,99 @@
+    --similar $similar
+    --max_mismatch $max_mismatch
+    --min_pairs $min_pairs
+    #if str( $other ) != "None":
```
I think you can just use `#if $other:`.

bernt-matthias replied:
done.

bgruening commented on tools/idba_ud/idba_ud.xml:
```diff
@@ -0,0 +1,99 @@
+
+    ;
+
+    mv out/scaffold.fa '$output'
```
you can remove this and use `from_work_dir`

bernt-matthias replied:
good idea. thanks

bgruening commented on tools/idba_ud/idba_ud.xml:
```diff
@@ -0,0 +1,99 @@
+
+    <inputs>
+        <param name="read" type="data" format="fasta" argument="--read" label="Fasta read file. Lower or equal to 250b"/>
+        <param name="long_read" type="data" format="fasta" optional="true" argument="--long_read" label="Fasta long read file. More than 250b"/>
```
if you use `arguement` in this way, you can remove name, if you like.

bernt-matthias replied:
cool. done.

bernt-matthias commented:
Also started to document how to generate interleaved fasta. Question remains: how to remove 'N's and do we want users to execute 2-3 steps? Wrt. testing these steps: one of our users who requested idba could test it right away.

bgruening commented:
> And another question: can/should we add the galaxy tools to the requirements?
What? :)

bernt-matthias commented:
OK. I removed fq2fa and updated the help (is there a nice way for rendering the links?). I did not mention the potential problem of non-standard bases .. hoping that they will be removed by quality filtering anyway.

nsoranzo commented on tools/idba_ud/idba_ud.xml:
```diff
@@ -0,0 +1,107 @@
+* *Samtools extract FASTA or FASTQ from a SAM file* (https://toolshed.g2.bx.psu.edu/view/devteam/fastq_to_fasta)
+
+Note that, IDBA-UD assumes that the paired-end reads are in order (-&gt;, &lt;-). 
+If your data is in reverse order (&lt;-, -&gt;), please convert it by yourself.
```
If you [CDATA](https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#help-tag) you don't need to use HTML entities like `&gt;`

bernt-matthias replied:
done

nsoranzo commented on tools/idba_ud/idba_ud.xml:
```diff
@@ -0,0 +1,107 @@
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements"/>
+    <command>
```
CDATA

nsoranzo commented on tools/idba_ud/idba_ud.xml:
```diff
@@ -0,0 +1,107 @@
+FASTA using the tools:  
+
+* *FASTQ interlacer on paired end read* (https://toolshed.g2.bx.psu.edu/view/devteam/fastq_paired_end_interlacer)
+* *Samtools extract FASTA or FASTQ from a SAM file* (https://toolshed.g2.bx.psu.edu/view/devteam/fastq_to_fasta)
```
The help section is interpreted as [reStructuredText](http://docutils.sourceforge.net/rst.html), so you can use the syntax for [embedded URIs](http://docutils.sourceforge.net/docs/user/rst/quickref.html#hyperlink-targets).

bernt-matthias replied:
done

bernt-matthias commented:
I fixed the last todos. Will meet a user next week who may tell me if the tool integrates nicely with the others as documented in the help. If so then the tool is ready to merge from my side.

bgruening commented:
@bernt-matthias can you ping us if you got the feedback or should we give this now a final review?

bernt-matthias commented:
Users are happy. Review is always a good idea .. if you have the time for it .. maybe diff against the original version.

bernt-matthias commented:
ready to merge from my side

bgruening commented:
Danke Matthias!
